### PR TITLE
PR 6187 Continue. Don't visit already checked successors

### PR DESCRIPTION
**Pull request by sklam**
- Build on #6187 
- Add test with the reproducer in #6186 and check the iteration count, which shouldn't change unless the algorithm is changed.
- Improved the fix in #6187. Use `set` and reduce iteration count in the backedge finding algorithm.

### CODE REVIEW

File: numba/tests/test_flow_control.py [hunk 1]
```diff
@@ -950,6 +950,79 @@ def test_loops(self):
...
+            758: {778},
+            778: {722},
+            816: {818},
+            818: set(),
+        })
+        g.set_entry_point(0)
+        g.process()
+        stats = {}
+        # Compute backedges and store the iteration count for testing
+        g._find_back_edges(stats=stats)
```
Comment (gmarkall) [inline numba/tests/test_flow_control.py:1023]: Would it be a good idea to also check the set of back edges?

```suggestion
        back_edges = g._find_back_edges(stats=stats)
        self.assertEqual(back_edges, {(666, 610), (778, 722)})
```

**Review by gmarkall — COMMENTED** [review]
The code changes look good to me. Just a question about the tests on the diff.

File: numba/tests/test_flow_control.py (hunk 1 above)
Comment (sklam) [reply numba/tests/test_flow_control.py:1023]: Good idea.

**Comment by sklam** [comment]
I have found a python source that will trigger this behavior:

```python
from numba import njit


@njit
def foo(a):
    x = 0
    if a < 1 or a < 2 or a < 3:
        x += 1
    else:
        x += 1

    if a < 1 or a < 2:
        x += 1

    if a < 1:

        if a < 1 or a < 2 or a < 3:
            x += 1
        else:
            x += 1

    else:

        if a < 1 or a < 2:
            x += 1

    if a < 1:
        x += 1

    if a < 1:
        x += 1

    if a < 1:
        x += 1

    if a < 1:
        x += 1

    if a < 1:
        x += 1

    for i in range(a):

        if a < 1 or a < 2 or a < 3:
            x += 1
        else:
            x += 1

        if a < 1 or a < 2:
            x += 1

        if a < 1:

            if a < 1 or a < 2 or a < 3:
                x += 1
            else:
                x += 1

    else:

        if a < 1 or a < 2 or a < 3:
            x += 1
        else:
            x += 1

        if a < 1 or a < 2:
            x += 1

        if a < 1:

            if a < 1 or a < 2 or a < 3:
                x += 1
            else:
                x += 1

        else:

            if a < 1 or a < 2:
                x += 1
    return x


foo(10)
```

File: numba/core/controlflow.py [hunk 2]
```diff
@@ -583,13 +588,17 @@ def _find_back_edges(self):
...
+
         def push_state(node):
             stack.append(node)
             succs_state[node] = [dest for dest in self._succs[node]]
 
         push_state(entry_point)
 
         while stack:
+            if stats is not None:
+                stats['iteration_count'] += 1
```
Comment (sklam) [inline numba/core/controlflow.py:601]: TODO: lift this branch out of the loop and just do a simple counter increment in the loop

**Review by gmarkall — APPROVED** [review]
This looks good to me!

**Review by stuartarchibald — APPROVED** [review]
Thanks for fixing this @gmarkall and @sklam. I've tested the backedge alg on mainline against the "pathological" case, takes about 11s locally, with this PR it takes 0.001s. Code looks good.

**Comment by stuartarchibald** [comment]
Tests are all passing on the checks page: https://github.com/numba/numba/pull/6195/checks it's reported incorrectly above.

**Pull request merged**